Migrate circuits card to view component 4958

Asrani-Aman opened the pull request:
Fixes #4958 

## Describe the changes you made in this PR

This PR introduces a new `CircuitCardComponent` to replace the existing inline card rendering for Editor Picks Circuit  and Recent  circuits. The changes include:

1. Creation of a new ViewComponent:
   - Added `app/component/circuit_card_component.rb`
   - Added `app/component/circuit_card_component.html.erb`

2. Refactoring of the circuit card logic:
   - Moved the card rendering logic from the view into the component
   - Encapsulated project image preview, name, and URL generation within the component

3. Implementation of `with_collection` for efficient rendering:
   - Modified the component to work with ViewComponent's `with_collection` method
   - Updated the views to use `with_collection` for rendering multiple project cards

4.  Made the card component reusable across different parts of the application

5. Utilized`project_image_preview` and `user_project_url` helpers into the component

6. Enhanced maintainability:
   - Centralized the card rendering logic, making future updates easier
   - Improved readability of both the component and the main views
   
7. Added unit tests for the new CircuitCardComponent

8. . **Added Preview with Notes and Params:**
   - Created a preview for the `CircuitCardComponent` to showcase how the component can be used with different parameters(screen recording is attached for the same).
   - The preview is defined in `spec/components/previews/circuit_card_component_preview.rb`.

These changes aim to improve code quality, maintainability, and performance in rendering project cards throughout the application.

### Screenshots of the changes (If any) -
![Screenshot 2024-07-06 at 12 52 23 AM](https://github.com/CircuitVerse/CircuitVerse/assets/96644946/3e0ac300-984c-4d76-b422-d82d7112d823)

![Screenshot 2024-07-06 at 1 42 30 AM](https://github.com/CircuitVerse/CircuitVerse/assets/96644946/b4f64800-e66c-473b-95bf-1b9f50f15608)

https://github.com/CircuitVerse/CircuitVerse/assets/96644946/3ad23251-52f6-488d-95c4-0fd2ae7b1ad1




Note: Please check **Allow edits from maintainers.** if you would like us to assist in the PR. 


<!-- This is an auto-generated comment: release notes by coderabbit.ai -->
## Summary by CodeRabbit

## Release Notes

- **New Features**
  - Introduced a responsive circuit card component displaying circuit details, including an image and a link for more information.
  - Enhanced the rendering of featured circuits and recent projects using a streamlined component-based approach.
  - Added a new footer component with social media and footer links, improving the overall layout.
  - Implemented email confirmation functionality for user registrations.

- **Bug Fixes**
  - Improved error handling in the circuit preview logic to prevent exceptions when accessing project details.

- **Tests**
  - Added a new RSpec test suite for the Circuit Card Component to ensure proper rendering and content display.
  - Introduced test suites for the Footer and Social Links components to validate rendering and user-specific link behavior.

- **Documentation**
  - Introduced preview classes for the Footer and Social Links components to facilitate component visualization during development.
<!-- end of auto-generated comment: release notes by coderabbit.ai -->

Asrani-Aman committed:
feat: add CircuitCard View Component

Asrani-Aman committed:
refactor: rendering CircuitCard view Component for optimisation of code

Asrani-Aman committed:
feat: add test case for CircuitCard View Component

Asrani-Aman committed:
fix: coldeclimate extra space space fix

Asrani-Aman committed:
feat: add Preview for CircuitCard Compoenent with notes and Params

Asrani-Aman committed:
Email Verification at the time of  user registration(#4957) (#4993)

dependabot[bot] committed:
build(deps): bump rexml from 3.2.7 to 3.3.2 (#5020)

Bumps [rexml](https://github.com/ruby/rexml) from 3.2.7 to 3.3.2.
- [Release notes](https://github.com/ruby/rexml/releases)
- [Changelog](https://github.com/ruby/rexml/blob/master/NEWS.md)
- [Commits](https://github.com/ruby/rexml/compare/v3.2.7...v3.3.2)

---
updated-dependencies:
- dependency-name: rexml
  dependency-type: direct:production
...

Signed-off-by: dependabot[bot] <support@github.com>
Co-authored-by: dependabot[bot] <49699333+dependabot[bot]@users.noreply.github.com>

dependabot[bot] committed:
build(deps): bump rails_admin from 3.1.2 to 3.1.3 (#5013)

Bumps [rails_admin](https://github.com/sferik/rails_admin) from 3.1.2 to 3.1.3.
- [Changelog](https://github.com/railsadminteam/rails_admin/blob/v3.1.3/CHANGELOG.md)
- [Commits](https://github.com/sferik/rails_admin/compare/v3.1.2...v3.1.3)

---
updated-dependencies:
- dependency-name: rails_admin
  dependency-type: direct:production
...

Signed-off-by: dependabot[bot] <support@github.com>
Co-authored-by: dependabot[bot] <49699333+dependabot[bot]@users.noreply.github.com>

dependabot[bot] committed:
build(deps): bump rexml from 3.2.5 to 3.3.2 in /docs (#5027)

Bumps [rexml](https://github.com/ruby/rexml) from 3.2.5 to 3.3.2.
- [Release notes](https://github.com/ruby/rexml/releases)
- [Changelog](https://github.com/ruby/rexml/blob/master/NEWS.md)
- [Commits](https://github.com/ruby/rexml/compare/v3.2.5...v3.3.2)

---
updated-dependencies:
- dependency-name: rexml
  dependency-type: indirect
...

Signed-off-by: dependabot[bot] <support@github.com>
Co-authored-by: dependabot[bot] <49699333+dependabot[bot]@users.noreply.github.com>

dependabot[bot] committed:
build(deps): bump cv-frontend-vue from `8fa90f7` to `5f66af1` (#5026)

Bumps [cv-frontend-vue](https://github.com/CircuitVerse/cv-frontend-vue) from `8fa90f7` to `5f66af1`.
- [Commits](https://github.com/CircuitVerse/cv-frontend-vue/compare/8fa90f78aefe81ecd7dd803335864c7b82539050...5f66af11a31a6dc17035924bbb55e77dd9e0c27b)

---
updated-dependencies:
- dependency-name: cv-frontend-vue
  dependency-type: direct:production
...

Signed-off-by: dependabot[bot] <support@github.com>
Co-authored-by: dependabot[bot] <49699333+dependabot[bot]@users.noreply.github.com>

dependabot[bot] committed:
build(deps): bump actionpack from 7.0.8.1 to 7.0.8.4 (#5029)

Bumps [actionpack](https://github.com/rails/rails) from 7.0.8.1 to 7.0.8.4.
- [Release notes](https://github.com/rails/rails/releases)
- [Changelog](https://github.com/rails/rails/blob/v7.1.3.4/actionpack/CHANGELOG.md)
- [Commits](https://github.com/rails/rails/compare/v7.0.8.1...v7.0.8.4)

---
updated-dependencies:
- dependency-name: actionpack
  dependency-type: indirect
...

Signed-off-by: dependabot[bot] <support@github.com>
Co-authored-by: dependabot[bot] <49699333+dependabot[bot]@users.noreply.github.com>

tachyons committed:
Skip vite from dependabot update

dependabot[bot] committed:
build(deps): bump rexml from 3.3.2 to 3.3.3 (#5038)

Bumps [rexml](https://github.com/ruby/rexml) from 3.3.2 to 3.3.3.
- [Release notes](https://github.com/ruby/rexml/releases)
- [Changelog](https://github.com/ruby/rexml/blob/master/NEWS.md)
- [Commits](https://github.com/ruby/rexml/compare/v3.3.2...v3.3.3)

---
updated-dependencies:
- dependency-name: rexml
  dependency-type: direct:production
...

Signed-off-by: dependabot[bot] <support@github.com>
Co-authored-by: dependabot[bot] <49699333+dependabot[bot]@users.noreply.github.com>

Waishnav committed:
feat: update simple_discussion (#5031)

Asrani-Aman committed:
Resend confirmation email #4957 (#4994)

Asrani-Aman committed:
fix: rubocop issues for CircuitCardComponent

Asrani-Aman committed:
fix: codeclimate and hound issues in CircuitCardComponent

Asrani-Aman committed:
fix: indentation issues

Asrani-Aman committed:
SAML based SSO  (#4989)

* feat: add ruby-saml gem

* fix: devise_saml_authenticable and ruby_saml versions compatibility

* feat: add methods in saml_session_controller for handling SAML request

* fix: add saml_update_resource_hook to resolve create method override in devise_saml_authenticatable

* feat: add routes for SAML based SSO

* style: update UI for sso

* fix: improve csrf protection in saml_session_controller.rb

* fix: codeclimate issues in saml_session_controller.rb

* refactor: remove extra line used for debugging

* fix: add double quotes code climate

* fix: code climate indentation issue

* refactor: saml_session_controller.rb

* refactor: Update saml_sessions_controller.rb

* fix: add blank line after private method

* refactor: remove extra spaces in saml_sessions_controller.rb

* feat: remove ruby-saml as it is dependency of gem devise-saml-authenticable

* feat: add gemfile.lock

* feat: enabling flipper

* refactor: adding methods in  saml_sessions_controller.rb

* fix: remove explicitly called methods in saml_session_controller

* refator: Update Gemfile

* refactor: update gemfile.lock

---------

Co-authored-by: Aboobacker MK <aboobackervyd@gmail.com>

Asrani-Aman committed:
fix: make componenet independent of current user

smritigarg commented on spec/components/circuit_card_component_spec.rb:
```diff
@@ -0,0 +1,16 @@
+
+RSpec.describe CircuitCardComponent, type: :component do
+  let(:circuit) { create(:project) }
+
```
Wrap this in one `describe block` 

```
describe "Circuit Card component rendering" do
    before do
      sign_in FactoryBot.create(:user)
    end
    
  ## here goes the it block 
end

```

gr455 committed:
Do not check contention on downstream values (#4903)

We define an output node value to be upstream if it is given to
it by its parent element. All other values are downstream values.

In the case that a node gives an output node value (while the
parent does not drive the output value. e.g. Tristate with control
bit 0), and then the same node gives another value to the output
node, earlier it would cause contention since the output node is
said to [before the propagation from the other node] still have
the former value.

This is a downstream value and post this commit will be
overwritten. Contention will only be checked at upstream valued
output nodes.

Asrani-Aman committed:
Migrate footer partial into a view component 4958 (#5011)

tachyons-gitlab committed:
chore: add API link

Asrani-Aman committed:
fix: sign_in user context for testing a componenet

Asrani-Aman commented:
Hii @tachyons , I think we can merge this PR

naman79820 commented:
closing as completed for now
